lib/sshutils/sftp/sftp.go

```diff
@@ -0,0 +1,472 @@
+	}
+
+	c := &Config{
+		srcPaths: []string{src},
+		dstPath:  dst,
+		srcFS:    &remoteFS{},
+		dstFS:    &localFS{},
+		opts:     opts,
+	}
+	c.setDefaults()
+
+	return c, nil
+}
+
+// setDefaults sets default values
+func (c *Config) setDefaults() {
+	logger := c.Log
+	if logger == nil {
+		logger = log.StandardLogger()
+	}
+	c.Log = logger.WithFields(log.Fields{
+		trace.Component: "SFTP",
+		trace.ComponentFields: log.Fields{
+			"SrcPaths":      c.srcPaths,
+			"DstPath":       c.dstPath,
+			"Recursive":     c.opts.Recursive,
+			"PreserveAttrs": c.opts.PreserveAttrs,
+		},
+	})
+}
+
+// TransferFiles transfers files from the configured source paths to the
+// configured destination path over SFTP
+func (c *Config) TransferFiles(ctx context.Context, sshClient *ssh.Client) error {
+	sftpClient, err := sftp.NewClient(sshClient)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := c.initFS(ctx, sshClient, sftpClient); err != nil {
+		return trace.Wrap(err)
+	}
+
+	transferErr := c.transfer(ctx)
+	closeErr := sftpClient.Close()
+	if transferErr != nil {
+		return trace.Wrap(transferErr)
+	}
+
+	return trace.Wrap(closeErr)
+}
+
+// initFS ensures the source and destination filesystems are ready to transfer
+func (c *Config) initFS(ctx context.Context, sshClient *ssh.Client, client *sftp.Client) error {
+	var haveRemoteFS bool
+
+	srcFS, srcOK := c.srcFS.(*remoteFS)
+	if srcOK {
+		srcFS.c = client
+		haveRemoteFS = true
+	}
+	dstFS, dstOK := c.dstFS.(*remoteFS)
+	if dstOK {
+		dstFS.c = client
+		haveRemoteFS = true
+	}
+
+	if haveRemoteFS {
+		if c.getHomeDir == nil {
+			c.getHomeDir = func() (_ string, err error) {
+				// if home directory has already been cached, just return it
+				if c.homeDir != "" {
+					return c.homeDir, nil
+				}
+
+				c.homeDir, err = getRemoteHomeDir(sshClient)
+				return c.homeDir, err
+			}
+		}
+		return trace.Wrap(c.expandPaths(srcOK, dstOK))
+	}
+
+	return nil
+}
+
+func (c *Config) expandPaths(srcIsRemote, dstIsRemote bool) (err error) {
+	if srcIsRemote {
+		for i, srcPath := range c.srcPaths {
+			c.srcPaths[i], err = expandPath(srcPath, c.getHomeDir)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+	if dstIsRemote {
+		c.dstPath, err = expandPath(c.dstPath, c.getHomeDir)
+	}
+
+	return trace.Wrap(err)
+}
+
+func expandPath(path string, getHomeDir homeDirRetriever) (string, error) {
+	if !needsExpansion(path) {
+		return path, nil
+	}
+
+	homeDir, err := getHomeDir()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	// this is safe because we verified that all paths are non-empty
+	// in CreateUploadConfig/CreateDownloadConfig
+	return filepath.Join(homeDir, path[1:]), nil
+}
+
+// needsExpansion returns true if path is '~' or begins with '~/'
+func needsExpansion(path string) bool {
+	if len(path) == 1 {
+		return path == "~"
+	}
+
+	path = filepath.ToSlash(path)
+	return strings.HasPrefix(path, "~"+string(filepath.Separator))
```

Comment:
If this code is used in the client, `filepath.Separator` will fail on Windows. When you copy files on any platform, most likely, you will use Unix format `~/foo`, but on Windows, the separator is different `\`.
I'd just use `~/` to make it explicit.

Reply:
Hmm I think I might just check for `~/` and `~\` when on Windows just to be safe